File: benchmarks/compute/correlation/Zig/correlation.zig
const std = @import("std");

// ---------- 可选数据规模 ----------
const SMALL = 0;
const STANDARD = 1;
const LARGE = 2;

// 默认数据规模（可通过编译选项覆盖）
const DATASET = STANDARD;

const M = switch (DATASET) {
    SMALL => 50,
    LARGE => 2000,
    else => 500,
};

const N = switch (DATASET) {
    SMALL => 60,
    LARGE => 2500,
    else => 600,
};

// ---------- 初始化随机矩阵 ----------
fn initArray(m: usize, n: usize, data: *[M][N]f64, float_n: *f64, seed: u32) void {
    float_n.* = @as(f64, @floatFromInt(n));
    var rng = std.rand.DefaultPrng.init(seed);
    for (0..m) |i| {
        for (0..n) |j| {
            data[i][j] = rng.random().float(f64);
        }
    }
}

// ---------- 打印部分输出矩阵（防止被编译器优化） ----------
fn printMatrix(m: usize, symmat: *[M][M]f64) void {
    const limit = @min(8, m);
    for (0..limit) |i| {
        for (0..limit) |j| {
            std.debug.print("{d:.2} ", .{symmat[i][j]});
        }
        std.debug.print("\n", .{});
    }
}

// ---------- 主计算过程 ----------
fn computeCorrelation(
    m: usize,
    n: usize,
    float_n: f64,
    data: *[M][N]f64,
    symmat: *[M][M]f64,
    mean: *[M]f64,
    stddev: *[M]f64,
) void {
    const eps = 0.1;

    // 均值
    for (0..m) |j| {
        mean[j] = 0.0;
        for (0..n) |i| {
            mean[j] += data[j][i];
        }
        mean[j] /= float_n;
    }

    // 标准差
    for (0..m) |j| {
        stddev[j] = 0.0;
        for (0..n) |i| {
            const val = data[j][i] - mean[j];
            stddev[j] += val * val;
        }
        stddev[j] = @sqrt(stddev[j] / float_n);
        if (stddev[j] <= eps) stddev[j] = 1.0;
    }

    // 中心化并标准化
    for (0..m) |j| {
        for (0..n) |i| {
            data[j][i] -= mean[j];
            data[j][i] /= @sqrt(float_n) * stddev[j];
        }
    }

    // 相关矩阵
    for (0..m) |i| {
        symmat[i][i] = 1.0;
        for (i + 1..m) |j| {
            symmat[i][j] = 0.0;
            for (0..n) |k| {
                symmat[i][j] += data[i][k] * data[j][k];
            }
            symmat[j][i] = symmat[i][j];
        }
    }
}

// ---------- 主函数 ----------
pub fn main() !void {
    var float_n: f64 = undefined;
    var data: [M][N]f64 = undefined;
    var symmat: [M][M]f64 = undefined;
    var mean: [M]f64 = undefined;
    var stddev: [M]f64 = undefined;

    // WASI 兼容的命令行参数解析
    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    const args = try std.process.argsAlloc(allocator);
    const seed = if (args.len > 1) 
        try std.fmt.parseInt(u32, args[1], 10)
    else 
        42;

    initArray(M, N, &data, &float_n, seed);

    const start = std.time.nanoTimestamp();
    computeCorrelation(M, N, float_n, &data, &symmat, &mean, &stddev);
    const end = std.time.nanoTimestamp();

    const elapsed_seconds = @as(f64, @floatFromInt(end - start)) / 1e9;
    std.debug.print("Finished correlation calculation in {d:.3} seconds\n", .{elapsed_seconds});

    printMatrix(M, &symmat); // 打印前8x8子矩阵
}

File: benchmarks/compute/correlation/C/correlation.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>
#include <time.h>

#define SMALL 0
#define STANDARD 1
#define LARGE 2


#if !defined(DATASET)
#define DATASET STANDARD
#endif

#if DATASET == SMALL
#define M 50
#define N 60
#elif DATASET == LARGE
#define M 2000
#define N 2500
#else
#define M 500
#define N 600
#endif


void init_array(int m, int n, double data[m][n], double *float_n, unsigned int seed) {
    *float_n = (double)n;
    srand(seed);
    for (int i = 0; i < m; i++)
        for (int j = 0; j < n; j++)
            data[i][j] = ((double)rand() / RAND_MAX);
}


void print_matrix(int m, double symmat[m][m]) {
    for (int i = 0; i < m && i < 8; i++) {
        for (int j = 0; j < m && j < 8; j++) {
            printf("%0.2f ", symmat[i][j]);
        }
        printf("\n");
    }
}


void compute_correlation(int m, int n, double float_n,
                         double data[m][n],
                         double symmat[m][m],
                         double mean[m],
                         double stddev[m]) {
    double eps = 0.1;


    for (int j = 0; j < m; j++) {
        mean[j] = 0.0;
        for (int i = 0; i < n; i++)
            mean[j] += data[j][i];
        mean[j] /= float_n;
    }


    for (int j = 0; j < m; j++) {
        stddev[j] = 0.0;
        for (int i = 0; i < n; i++) {
            double val = data[j][i] - mean[j];
            stddev[j] += val * val;
        }
        stddev[j] = sqrt(stddev[j] / float_n);
        if (stddev[j] <= eps) stddev[j] = 1.0;
    }


    for (int j = 0; j < m; j++)
        for (int i = 0; i < n; i++) {
            data[j][i] -= mean[j];
            data[j][i] /= (sqrt(float_n) * stddev[j]);
        }


    for (int i = 0; i < m; i++) {
        symmat[i][i] = 1.0;
        for (int j = i + 1; j < m; j++) {
            symmat[i][j] = 0.0;
            for (int k = 0; k < n; k++)
                symmat[i][j] += data[i][k] * data[j][k];
            symmat[j][i] = symmat[i][j];
        }
    }
}

int main(int argc, char** argv) {
    double float_n;
    static double data[M][N];
    static double symmat[M][M];
    static double mean[M];
    static double stddev[M];

    unsigned int seed = (argc > 1) ? atoi(argv[1]) : 42;

    init_array(M, N, data, &float_n, seed);

    clock_t start = clock();
    compute_correlation(M, N, float_n, data, symmat, mean, stddev);
    clock_t end = clock();

    printf("Finished correlation calculation in %.3f seconds\n",
           (double)(end - start) / CLOCKS_PER_SEC);

    print_matrix(M, symmat);
    return 0;
}

